bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/RpcThrottler.java

```diff
@@ -172,13 +172,18 @@ public void awaitCompletion() throws InterruptedException {
 
         long now = clock.nanoTime();
         if (now >= noSuccessCheckDeadlineNanos) {
-          // There are unusual cases where an RPC could be completed, but we don't clean up
-          // the state and the locks.  Try to clean up if there is a timeout.
-          for (RetryHandler retryHandler : outstandingRetries.values()) {
-            retryHandler.performRetryIfStale();
-          }
-          if (isFlushed()) {
-            break;
+          lock.lock();
+          try {
+            // There are unusual cases where an RPC could be completed, but we don't clean up
+            // the state and the locks.  Try to clean up if there is a timeout.
+            for (RetryHandler retryHandler : outstandingRetries.values()) {
+              retryHandler.performRetryIfStale();
```

Comment:
I don't think this PR addresses the problem. There's an extra lock here but the lock was already acquired on line 168 and the lock is reentrant so I don't think this does anything. As best I can tell the problem is the call stack that can happen here (all while holding that lock):
`RetryHandler.performRetryIfStale`
`BulkMutation.Batch#setRetryComplete`
`RpcThrottler#onRetryCompletion`

At this point `outstandingRetries.remove(id)` is called, which will cause a `ConcurrentModificationException` on the next iteration on this line.

If that's all true, then we should make a local copy of `oustandingRetries.values()` (because it's backed by the map: https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#values()) and iterate over that.